### PR TITLE
User list export: apply filters (fixes #298)

### LINKED ISSUE #298: User list export - Filter by admission year, workshop, role (műhelytitkár)
Enable exporting only those users' data who were filtered (by admission year, workshop, role (műhelytitkár), etc.)

### PR DESCRIPTION
**Pull request by Trigary**
Summary of changes:

- The implementation behind the export button on the `/users` route has been replaced
  - The rendering has been moved from `list.blade.php` to `list_users_component.blade.php`
  - Clicking the button no longer calls the `users.export` route: the click is now handled within the livewire component
  - The same filtering logic is used within the component for the display of the users and for the exporting
- ~~To implement the filtering, a `UserFilter` utility class has been created~~
- The different `Export` classes under `UsersSheets` have been modified so that the included users can be specified
- The behavior of the `users.export` path (hopefully) hasn't changed: the same users will get included in the exports as before

Small, mostly unrelated fixes:

- The `$show_feedback` property has been deleted from `SemesterEvaluationExport`: it was unused
  - `UsersExport` also assumed that `SemesterEvaluationExport` had a `boolean` constructor parameter, while in fact, it did not have any parameters. This has also been fixed.
- `ListUsers`'s status related methods incorrectly named their parameters `status_id`, while in reality they received a status (string), not an ID

### CODE REVIEW

File: app/Utils/UserFilter.php [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+ *
+ * Usage:
+ * <ol>
+ * <li>Create a user query that should be further narrowed down, e.g. {@see User::canView()}</li>
+ * <li>Create a new class instance with the previously obtained query instance</li>
+ * <li>Call the desired filter methods on the instance to narrow down the query</li>
+ * <li>Get the final query by calling {@see UserFilter::getQuery()} on the instance</li>
+ * </ol>
+ */
+class UserFilter
```
Comment (kdmnk) [inline app/Utils/UserFilter.php]: The functionality of this class can be achieved by Laravel's scopes (see https://laravel.com/docs/10.x/eloquent#local-scopes). Rewrite it with scopes and try to create as general scopes as possible.

File: app/Http/Controllers/Secretariat/UserController.php [hunk 2]
```diff
@@ -302,7 +302,7 @@ public function export()
     {
         $this->authorize('viewAny', User::class);
 
-        return Excel::download(new UsersExport(), 'uran_export.xlsx');
+        return Excel::download(new UsersExport(User::canView()), 'uran_export.xlsx');
```
Comment (kdmnk) [inline app/Http/Controllers/Secretariat/UserController.php]: You don't use this controller function anymore. Delete it together with it's endpoint.

File: app/Exports/UsersExport.php [hunk 3]
```diff
@@ -7,20 +7,32 @@
...
+use Illuminate\Database\Eloquent\Builder;
 use Maatwebsite\Excel\Concerns\WithDefaultStyles;
 use Maatwebsite\Excel\Concerns\WithMultipleSheets;
 use PhpOffice\PhpSpreadsheet\Style\Style;
 
 class UsersExport implements WithMultipleSheets, WithDefaultStyles
 {
+    private Builder|User $includedUsersQuery;
+
+    public function __construct(Builder|User $includedUsersQuery)
```
Comment (kdmnk) [inline app/Exports/UsersExport.php]: why don't you pass the User collection? Using builders here makes it more complex. Order the query by name beforehand and you will have the same result.

File: app/Exports/UsersExport.php [hunk 4]
```diff
@@ -7,20 +7,32 @@
...
+    {
+        $this->includedUsersQuery = $includedUsersQuery;
+    }
+
     public function sheets(): array
     {
-        $sheets = [new CollegistsExport(), new StatusesExport()];
+        $sheets = [
+            new CollegistsExport($this->includedUsersQuery->clone()),
```
Comment (kdmnk) [inline app/Exports/UsersExport.php]: why do you need `clone`?

File: resources/views/secretariat/user/list_users_component.blade.php [hunk 5]
```diff
@@ -123,4 +123,11 @@
             </div>
         </div>
     </div>
+
+    {{-- Export --}}
+    <div class="fixed-action-btn tooltipped" data-position="left" data-tooltip="Exportálás">
```
Comment (kdmnk) [inline resources/views/secretariat/user/list_users_component.blade.php]: Make it explicit: eg. "szűrt felhasználók adatainak exportálása"

**Review by kdmnk — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/Secretariat/UserController.php (hunk 2 above)
Comment (Trigary) [reply app/Http/Controllers/Secretariat/UserController.php]: I think it's still in use: it's referenced in `evaluation_form_available_details.blade.php` and in `evaluation_form_closed.blade.php`. (Look for `<a href="{{ route('users.export') }}">letölthetőek</a>` in both files.)

File: app/Exports/UsersExport.php (hunk 4 above)
Comment (Trigary) [reply app/Exports/UsersExport.php]: Different classes use the query passed to them in differing ways. By cloning, we ensure that modifications in one export class don't modify the behavior of the consequent exports.

If we were to pass a collection instead of a builder, then cloning would not be necessary. I'll use a collection.

File: app/Exports/UsersExport.php (hunk 3 above)
Comment (Trigary) [reply app/Exports/UsersExport.php]: Done!

File: app/Utils/UserFilter.php (hunk 1 above)
Comment (Trigary) [reply app/Utils/UserFilter.php]: Thanks!

File: app/Http/Controllers/Secretariat/UserController.php (hunk 2 above)
Comment (kdmnk) [reply app/Http/Controllers/Secretariat/UserController.php]: Hm, true but it might create some confusion in the future. These are only in emails, we could change the link to point to the users index page.

Comment (Trigary) [reply app/Http/Controllers/Secretariat/UserController.php]: Done

File: app/Models/User.php [hunk 6]
```diff
@@ -589,6 +589,65 @@ public function scopeHasToPayKKTNetregInSemester(Builder $query, int $semester_i
...
+     * The roles are specified by their IDs.
+     */
+    public function scopeHasAllRoleIds(Builder $query, array $roleIdsAll): void
+    {
+        foreach ($roleIdsAll as $roleId) {
+            $query->whereHas('roles', function (Builder $query) use ($roleId) {
+                $query->where('id', $roleId);
+            });
+        }
+    }
```
Comment (kdmnk) [inline app/Models/User.php]: Please move this to App\Utils\HasRole trait. 
If you want, you could also investigate if it's possible to chain the `withRole` scope for multiple roles. Please crate some test cases if so. If not, it's fine for now. I guess it won't result in an optimal query anyways.

File: app/Models/User.php [hunk 7]
```diff
@@ -589,6 +589,65 @@ public function scopeHasToPayKKTNetregInSemester(Builder $query, int $semester_i
...
+
+    /**
+     * Scope a query to only include users who are in all the specified workshops.
+     * The workshops are specified by their IDs.
+     */
+    public function scopeInAllWorkshopIds(Builder $query, array $workshopsIdsAll): void
+    {
+        foreach ($workshopsIdsAll as $workshopId) {
+            $query->whereHas('workshops', function (Builder $query) use ($workshopId) {
+                $query->where('id', $workshopId);
```
Comment (kdmnk) [inline app/Models/User.php]: use whereIn and you can pass the $workshopIdsAll without the foreach

File: app/Models/User.php [hunk 8]
```diff
@@ -589,6 +589,65 @@ public function scopeHasToPayKKTNetregInSemester(Builder $query, int $semester_i
...
+
+    /**
+     * Scope a query to only include users who have all the specified roles.
+     * The roles are specified by their IDs.
+     */
+    public function scopeHasAllRoleIds(Builder $query, array $roleIdsAll): void
+    {
+        foreach ($roleIdsAll as $roleId) {
+            $query->whereHas('roles', function (Builder $query) use ($roleId) {
+                $query->where('id', $roleId);
```
Comment (kdmnk) [inline app/Models/User.php]: use whereIn and you can pass the $roleIdsAll without the foreach. Check the documentation to make sure that the behaviour won't be 'hasAnyRoleIds'...

**Review by kdmnk — CHANGES_REQUESTED** [review]

File: app/Models/User.php (hunk 8 above)
Comment (Trigary) [reply app/Models/User.php]: Done! The documentation didn't help to be honest, but maybe I just didn't find the correct documentation. Testing suggests the solution indeed works.

File: app/Models/User.php (hunk 7 above)
Comment (Trigary) [reply app/Models/User.php]: Done!

File: app/Models/User.php (hunk 6 above)
Comment (Trigary) [reply app/Models/User.php]: Done!

**Comment by Trigary** [comment]
The new `scopeWithAllRoles` doesn't (yet) support specifying role objects and the new tests don't use role objects. Let me know if I should change that.

**Comment by kdmnk** [comment]
> The new `scopeWithAllRoles` doesn't (yet) support specifying role objects and the new tests don't use role objects. Let me know if I should change that.

Yeah I was also thinking about that but honestly I kind of want to rewrite the whole scope to support multiple roles in an optimal way and then use the scope for the hasRole function.

Whether to support object arguments or IDs is also a debatable. Objects improve readability and static typing whereas IDs are usually more optimal (if you only have the id you should be able to pass it as it is).

In short, it's fine by me, but feel free to investigate it further.

I'll take a look at this `whereHas` use case later, seems strange.

**Comment by kdmnk** [comment]
FYI you can use https://laravel.com/docs/11.x/queries#debugging to see make sure the queries are good. Or also in laravel debugbar.

**Review by kdmnk — APPROVED** [review]
Checked the docs. LGTM, if you want to merge now.

**Comment by Trigary** [comment]
I want to merge now.
